### ISSUE TITLE
Reader model inference - passage score confusing the reader model?

### ISSUE THREAD

**Issue by jobergum** [facebookresearch/DPR#45]
Hello! 

First, thanks for sharing this work! 

I got a setup where I have a successfully reproduced the DPR retriever accuracy at top 20 and top 100 retrieved passages using vespa.ai but I have some trouble with the reader checkpoint and the EM metric/results it produces in my setup and wanted to ask if you have any idea. I suspect it's due to the score range distribution. With Vespa.ai the passage score is calculated as 1/(1+l2_distance) (After transforming the inner dot product space to euclidian space for HNSW). Steps to reproduce with two questions in the NQ open test set

```
python3 data/download_data.py --resource  checkpoint.reader.nq-single.hf-bert-base
Loading from  https://dl.fbaipublicfiles.com/dpr/checkpoint/reader/nq-single/hf_bert_base.cp
100% [....................................................................] 1313930953 / 1313930953Saved to  ./checkpoint/reader/nq-single/hf-bert-base.cp
```

**Reader model inference**

```
 python3 train_reader.py --prediction_results_file out.json  --eval_top_docs 10 100 --model_file data/checkpoint/reader/nq-single-subset/hf-bert-base.cp --dev_file results-filtered.json --passages_per_question_predict 100 --sequence_length 350
```

Where ``results-filtered.json`` is available https://gist.githubusercontent.com/jobergum/33f6fd7f612dbb923e1414682fc66f5b/raw/9be34f2518c5d9ef132fa60f0e7689c5f724ec13/results-filtered.json. Variants of those two questions exists in the train set (https://arxiv.org/pdf/2008.02637.pdf) so I was hoping for a good EM score.

**Reader inference Output**
```
on device=cpu, n_gpu=0, world size=1
16-bits training: False 
 **************** CONFIGURATION **************** 
adam_betas                     -->   (0.9, 0.999)
adam_eps                       -->   1e-08
batch_size                     -->   2
checkpoint_file_name           -->   dpr_reader
dev_batch_size                 -->   4
dev_file                       -->   results-filtered.json
device                         -->   cpu
distributed_world_size         -->   1
do_lower_case                  -->   False
dropout                        -->   0.1
encoder_model_type             -->   None
eval_step                      -->   2000
eval_top_docs                  -->   [10, 100]
fp16                           -->   False
fp16_opt_level                 -->   O1
fully_resumable                -->   False
gold_passages_src              -->   None
gold_passages_src_dev          -->   None
gradient_accumulation_steps    -->   1
learning_rate                  -->   1e-05
local_rank                     -->   -1
log_batch_step                 -->   100
max_answer_length              -->   10
max_grad_norm                  -->   1.0
max_n_answers                  -->   10
model_file                     -->   data/checkpoint/reader/nq-single-subset/hf-bert-base.cp
n_gpu                          -->   0
no_cuda                        -->   False
num_train_epochs               -->   3.0
num_workers                    -->   16
output_dir                     -->   None
passages_per_question          -->   2
passages_per_question_predict  -->   100
prediction_results_file        -->   out.json
pretrained_file                -->   None
pretrained_model_cfg           -->   None
projection_dim                 -->   0
seed                           -->   0
sequence_length                -->   350
train_file                     -->   None
train_rolling_loss_step        -->   100
warmup_steps                   -->   100
weight_decay                   -->   0.0
 **************** CONFIGURATION **************** 
***** Initializing components for training *****
Reading saved model from data/checkpoint/reader/nq-single-subset/hf-bert-base.cp
model_state_dict keys odict_keys(['model_dict', 'optimizer_dict', 'scheduler_dict', 'offset', 'epoch', 'encoder_params'])
Overriding args parameter value from checkpoint state. Param = pretrained_model_cfg, value = bert-base-uncased
Overriding args parameter value from checkpoint state. Param = encoder_model_type, value = hf_bert
Overriding args parameter value from checkpoint state. Param = sequence_length, value = 350
..
Loading checkpoint @ batch=3940 and epoch=19
Loading model weights from saved state ...
Loading saved optimizer state ...
No train files are specified. Run validation.
Validation ...
Data files: ['results-filtered.json']
Found preprocessed files. ['results-filtered.0.pkl', 'results-filtered.1.pkl']
Reading file results-filtered.0.pkl
Aggregated data size: 1
Reading file results-filtered.1.pkl
Aggregated data size: 2
Total data size: 2
n=10	EM 0.00
n=100	EM 0.00
Saving prediction results to  out.json
```

**out.json**
```json
[
    {
        "gold_answers": [
            "Jakeem Grant",
            "John Ross"
        ],
        "predictions": [
            {
                "prediction": {
                    "passage": "4. 18 run by within the same week added some support to the legitimacy of the times.'s was hand - timed by a scout as running a 4. 10 in 2016, potentially beating's record. ran a 4. 27 - second 40 - yard dash in 1989. 2013, recorded a time of 4. 22 at a facility during a workout. 2017 sprinter ran a time of 4. 12 seconds on turf in response to claims that players are as fast as. is a list of the official 40 - yard",
                    "passage_idx": 0,
                    "relevance_score": -3.989269256591797,
                    "score": -5.150297403335571,
                    "text": "scout"
                },
                "top_k": 10
            },
            {
                "prediction": {
                    "passage": "was a high school track champion and record holder in the late 1940s. was a standout sprinter for in,.'s speed was such that the athletic director steered him toward track, not wanting him to get hurt playing football. soon became a statewide success and was followed by the state's largest newspaper, the \" \". ran the 40 yard dash in 4. 2 seconds, a schoolboy record at the time, the 100 yard dash in a wind - assisted 9. 2 seconds, and the 220 yard sprint",
                    "passage_idx": 28,
                    "relevance_score": -2.3115363121032715,
                    "score": 8.016469955444336,
                    "text": ""
                },
                "top_k": 100
            }
        ],
        "question": "who ran the fastest 40 yard dash in the nfl"
    },
    {
        "gold_answers": [
            "Kevin McKeon as Young Pink",
            "Bob Geldof as Pink",
            "David Bingham as Little Pink",
            "Bob Geldof",
            "David Bingham",
            "Kevin McKeon"
        ],
        "predictions": [
            {
                "prediction": {
                    "passage": "the tour, band relationships dropped to an all - time low ; four were parked in a circle, with the doors facing away from the centre. used his own vehicle to arrive at the venue, and stayed in separate hotels from the rest of the band., returning to perform his duties as a salaried musician, was the only member of the band to profit from the tour, which lost about \u00a3400, 000. film adaptation, \" \u2013, \" was released in 1982. was written by and directed by, with as.",
                    "passage_idx": 2,
                    "relevance_score": -11.945453643798828,
                    "score": -4.854901194572449,
                    "text": "band"
                },
                "top_k": 10
            },
            {
                "prediction": {
                    "passage": "\" \", alongside and in 2016. - is married to. first child, a son, was born in 2015. has a brother and sister, &. grew up on the cost of. 2016, - revealed that shortly after relocating to in 2012, he began receiving treatment from. actor told of \" \" that he decided to attend because he did not like the personality changes he suffered as a result of consuming alcohol. -.",
                    "passage_idx": 32,
                    "relevance_score": -7.610385417938232,
                    "score": 7.239379405975342,
                    "text": ""
                },
                "top_k": 100
            }
        ],
        "question": "who played pink in pink floyd the wall"
    }
]
```

**Comment by vlad-karpukhin** [comment]
Hello,

Thanks for your interest.

1. The retriever scores don't participate in our reader model inference process (although they are presented in the results file, that is just for future experiments). So different scores in Vespa should't be an issue.

2. You will need to use checkpoint.reader.nq-single.hf-bert-base as the reader checkpoint, not checkpoint.reader.nq-single-subset.hf-bert-base.

3. Empty predicted spans for top 100 predictions seems strange.

So far, seems like some DPR-internal issue.

Let me know if you will have the same results with the recommended reader checkpoint and will try to reproduce it.

**Comment by jobergum** [comment]
Thanks a lot. Know I get the expected result with EM 100.

**Comment by shmsw25** [comment]
Hi @jobergum, just to clarify, were you able to reproduce [the result (34.8 EM)](https://efficientqa.github.io/getting_started.html) with checkpoint.reader.nq-single-subset.hf-bert-base? I wonder if you were not able to reproduce, or you did reproduce but was just wondering why it doesn't answer to some questions that appear in the train set.

**Comment by jobergum** [comment]
@shmsw25 I'm using the `checkpoint/reader/nq-single/hf-bert-base.cp` model as suggested by @vlad-karpukhin I get results more inline with the reported EM in the paper. 

With it I get 
**Reader EM**
```
n=10    EM 38.81
n=100   EM 39.00
```
**Retriever Recall**
```
Recall@100 0.836011 
Recall@20 0.778393 
Recall@10 0.727978 
MRR@100 0.545895 
```

With `checkpoint.reader.nq-single-subset.hf-bert-base`` with the same retrieved set I got a low EM score and I used those two questions to debug as similar question occurs in both test and train.
